raysect/core/math/function/function2d/base.pyx

```diff
@@ -724,3 +725,16 @@ cdef class GreaterEqualsScalar2D(Function2D):
 
     cdef double evaluate(self, double x, double y) except? -1e999:
         return self._value >= self._function.evaluate(x, y)
+
+
+cdef bint is_callable(object f):
+    """
+    Tests if an object is a python callable or Function2D object.
+
+    :param object f: Object to test.
+    :return: True if callable, False otherwise.
+    """
+    # VectorFunction2D and Function2D objects are not equivalent.
+    if isinstance(f, VectorFunction2D):
```

Comment:
The reason for this is that `VectorFunction2D` has a `__call__` method, and so `callable(f)` will return `True` if `f` is a `VectorFunction2D` instance. This prevents the vector arithmetic methods (e.g. f2d + vf2d) being used when one of the operands is a `Function2D`: the `is_callable` check in e.g. `Function2D.__add__` returns true and an `AddFunction2D` object is returned, with the vector function incorrectly wrapped in a `PythonFunction2D` object.

What needs to happen is that the Function2D `is_callable` method returns `False` for `VectorFunction2D`. This then allows the `Function2D.__add__` method to fall through to the end and return `NotImplemented`, so that the `VectorFunction2D.__add__` method can be tried instead. This will then correctly return an `AddVectorFunction2D` object.

I encountered this when writing the arithmetic unit tests for the vector function. I didn't include the explanation for the change in `is_callable` because I thought it was a bit too long winded. Is it worth adding the explanation as a comment inside the function?

Reply:
This is one of the issues I was aware could occur when we headed down the path of performing operations between arbitrary function objects, hence my initial nervousness of exploring them. I think the only way we can support this in a scalable way is introduce a base class for the function framework. E.g. `class Function` then base classes for the return types: `ScalarFunction(Function)`, `Vector3DFunction(Function)` so it is possible to differentiate between python and function framework classes and handle or, defer handling them as appropriate. We could have new methods to keep everything literate, rather than endless isinstance(x, y):

```
is_raysect_function()
is_python_function()
```

Btw, brief aside. It could be a complete arse, but I really starting to think we may need to rename some of these classes , especially if we introduce more return types:

```
raysect.core.math.function.Function
    raysect.core.math.function.ScalarFunction
        raysect.core.math.function.ScalarFunction1D
        raysect.core.math.function.ScalarFunction2D
        raysect.core.math.function.ScalarFunction3D
    raysect.core.math.function.BoolFunction
        raysect.core.math.function.BoolFunction1D
        raysect.core.math.function.BoolFunction2D
        raysect.core.math.function.BoolFunction3D
    raysect.core.math.function.Vector2DFunction
        raysect.core.math.function.Vector2DFunction1D
        raysect.core.math.function.Vector2DFunction2D
        raysect.core.math.function.Vector2DFunction3D
    raysect.core.math.function.Vector3DFunction
        raysect.core.math.function.Vector3DFunction1D
        raysect.core.math.function.Vector3DFunction2D
        raysect.core.math.function.Vector3DFunction3D
```
Something for a later release perhaps.

Reply:
The `FunctionxD` classes are already spread quite widely throughout the code, both in Raysect and also in downstream packages such as Cherab and its submodules. Renaming these to `ScalarFunctionxD` would result in quite a lot of code churn, and I'm not sure it would gain much.

I'm not convinced of the need for `BoolFunctionxD`, given the casting rules which enable numbers to be converted to bool. The only reason the `VectorFunctionxD` classes exist is because it's not possible any other way to have a fast, cythonised evaluation of an expression returning a `Vector3D` object.

I do think that renaming the vector functions to `Vector3DFunctionxD` is a good idea though: it was a bit confusing for me initially when I saw `VectorFunction2D` was returning a `Vector3D` rather than a `Vector2D`. It does as you say also open up the possibility of adding `Vector2DFunctionxD` classes in future, but I'd rather wait until there was a concrete use case for these before going ahead and implementing them (even though the implementation would mostly be `sed 's/Vector3D/Vector2D/'`).